src/BinSkim.Driver/RuleTelemetryLogger.cs

```diff
@@ -0,0 +1,179 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+
+using Microsoft.ApplicationInsights;
+using Microsoft.ApplicationInsights.DataContracts;
+using Microsoft.CodeAnalysis.Sarif;
+
+namespace Microsoft.CodeAnalysis.IL
+{
+    /// <summary>
+    /// A logger that records rule summary telemetry by aggregating result kinds for each rule.
+    /// </summary>
+    internal sealed class RuleTelemetryLogger : IAnalysisLogger
+    {
+        // Application Insights event names
+        internal const string RuleSummaryEventName = "RuleSummary";
+        internal const string AnalysisStoppedEventName = "AnalysisStopped";
+
+        // Application Insights property names
+        internal const string RuleIdPropertyName = "RuleId";
+        internal const string RuntimeConditionsPropertyName = "Flags";
+
+        // Application Insights metric names
+        internal const string PassCountMetricName = "Pass";
+        internal const string FailCountMetricName = "Fail";
+        internal const string OpenCountMetricName = "Open";
+        internal const string ReviewCountMetricName = "Review";
+        internal const string NotApplicableCountMetricName = "NA";
+        internal const string InformationalCountMetricName = "Info";
+
+        /// <summary>
+        /// The Application Insights telemetry client.
+        /// </summary>
+        private readonly TelemetryClient telemetryClient;
+
+        /// <summary>
+        /// Dictionary of aggregated counts of result kinds keyed on the rule ID.
+        /// </summary>
+        private readonly Dictionary<string, ResultKindCounts> metricsMap = new Dictionary<string, ResultKindCounts>(StringComparer.Ordinal);
+
+        /// <summary>
+        /// Construct a new <see cref="RuleTelemetryLogger"/>.
+        /// </summary>
+        /// <param name="telemetryClient">The Application Insights telemetry client.</param>
+        public RuleTelemetryLogger(TelemetryClient telemetryClient)
+        {
+            this.telemetryClient = telemetryClient ?? throw new ArgumentNullException(nameof(telemetryClient));
+        }
+
+        public void AnalysisStarted()
```

Comment:
>AnalysisStarted [](http://example.com/codeflow?start=20&length=15)

So, no start event? I can appreciate you have sketched out some (very useful) things here and we may want to defer building out additional data not immediately relevant to our immediate goal to measure occurrences for specific rules. Eventually, I think perhaps a timestamp at start and analysis stopped would be useful, to measure how long binskim takes to analyze in various contexts.

Reply:
Raises an interesting point. In App Insights, something with a duration is known as an Operation. There are two types of Operation: Requests and Dependencies. It would be natural to model the operation of "Analysis" as a "Request" with a start/stop. The nice thing about that, is that you emit just a single event when the operation ends with the duration. It'll be logged as a "RequestTelemetry" with a timestamp (the stopping time) and a duration from which you can compute the start time, if necessary.

Reply:
Nice! It makes sense that the AI interaction itself can generate good timestamps for kick-off; there's no need for explicit activity here in the `AnalysisStarted` callback...

Reply:
OK. I implemented that in the latest iteration (11).
One little wrinkle: I had to call AnalysisStarted manually because it's called before we get a chance to add this logger. Might want to look at deferring the AnalysisStarted call in MultithreadedAnalyzeCommandBase.Analyze until after CreateContext. Or have another callback mechanism to let derived classes customize the logger.